mdn/src/main/scala/org/apache/james/mdn/MDNReportParser.scala

```diff
@@ -0,0 +1,475 @@
+      }
+      val sending = sendingMode match {
+        case "MDN-sent-manually" => DispositionSendingMode.Manual
+        case "MDN-sent-automatically" => DispositionSendingMode.Automatic
+      }
+      (action, sending)
+    })
+  }
+
+  //    action-mode = "manual-action" / "automatic-action"
+  private def actionMode = rule { "manual-action" | "automatic-action" }
+
+  //    sending-mode = "MDN-sent-manually" / "MDN-sent-automatically"
+  private def sendingMode = rule {"MDN-sent-manually" | "MDN-sent-automatically" }
+
+  /*    disposition-type = "displayed" / "deleted" / "dispatched" /
+                      "processed"    */
+  private def dispositionType : Rule1[DispositionType] = rule {
+    "displayed" ~ push(DispositionType.Displayed) |
+    "deleted" ~ push(DispositionType.Deleted) |
+    "dispatched" ~ push(DispositionType.Dispatched) |
+    "processed" ~ push(DispositionType.Processed)
+  }
+
+  private def dispositionModifiers: Rule1[Seq[DispositionModifier]] = rule { (ows ~ "/" ~ ows ~ capture(dispositionModifier) ~
+      zeroOrMore(ows ~ "," ~ ows ~ capture(dispositionModifier))) ~> ((head: String, tail: Seq[String]) =>
+      tail.prepended(head).map(new DispositionModifier(_)))
+    }
+
+
+  //    disposition-modifier = "error" / disposition-modifier-extension
+  private def dispositionModifier = rule { "error" | dispositionModifierExtension }
+
+  //    disposition-modifier-extension = Atom
+  private def dispositionModifierExtension = rule { atom }
+
+  //    error-field = "Error" ":" *([FWS] text)
+  private[mdn] def errorField: Rule1[Error] = rule { ("Error" ~ ":" ~ capture(zeroOrMore(optional(fws) ~ text))) ~> ((error: String) =>  new Error(Text.fromRawText(error))) }
+
+  //    extension-field = extension-field-name ":" *([FWS] text)
+  private[mdn] def extentionField: Rule1[ExtensionField] = rule { capture(extensionFieldName) ~ ":" ~ capture(zeroOrMore(optional(fws) ~ text)) ~> ((extensionFieldName: String, text : String) =>
+    ExtensionField.builder()
+      .fieldName(extensionFieldName)
+      .rawValue(text)
+      .build()) }
+
+  //    extension-field-name = field-name
+  private def extensionFieldName: Rule0 = rule { fieldName }
+
+  //   field-name      =   1*ftext
+  private def fieldName: Rule0 = rule { oneOrMore(ftext) }
+
+  /*   ftext           =   %d33-57 /          ; Printable US-ASCII
+                         %d59-126           ;  characters not including
+                                            ;  ":".   */
+  private def ftext: Rule0 = rule {
+    CharPredicate(33.toChar to 57.toChar) |
+    CharPredicate(59.toChar to 126.toChar)
+  }
+
+  //   CFWS            =   (1*([FWS] comment) [FWS]) / FWS
+  private def cfws: Rule0 = rule { (oneOrMore(optional(fws) ~ comment) ~ fws) | fws }
+
+  //   FWS             =   ([*WSP CRLF] 1*WSP) /  obs-FWS
+  private def fws: Rule0 = rule { (optional(zeroOrMore(wsp) ~ crlf) ~ oneOrMore(wsp)) | obsFWS }
```

Comment:
you are using optional instead of `?`, is there any reason?

Reply:
it is exactly the same thing, i will uniformize this